File: app/src/main/java/com/kirawii/thunderswufe/ml/ElectricityPredictor.java
package com.kirawii.thunderswufe.ml;

import android.content.Context;
import android.util.Log;
import com.kirawii.thunderswufe.data.database.ElectricityRecord;
import org.json.JSONArray;
import org.json.JSONObject;
import org.tensorflow.lite.Interpreter;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.channels.FileChannel;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class ElectricityPredictor {
    private static final String LSTM_MODEL_FILE = "lstm_model.tflite";
    private static final String LR_KERAS_MODEL_FILE = "linear_regression_keras.tflite";
    private static final String SCALER_PARAMS_FILE = "scaler_params.json";
    private static final int DEFAULT_INPUT_LENGTH = 7;
    private static final int MODEL_OUTPUT_STEPS = 1;
    private static final int PREDICTION_HORIZON_DAYS = 3;
    private static final float FLOAT_COMPARISON_THRESHOLD = 1e-6f;

    private final Context context;
    private Interpreter lstmInterpreter;
    private Interpreter lrKerasInterpreter;
    private float[] scalerScale;
    private float[] scalerMin;

    public ElectricityPredictor(Context context) {
        this.context = context;
        loadAllModelsAndParams();
    }

    private void loadAllModelsAndParams() {
        lstmInterpreter = loadTFLiteModel(LSTM_MODEL_FILE);
        if (lstmInterpreter != null) {
            Log.i("ElectricityPredictor", "LSTM model '" + LSTM_MODEL_FILE + "' loaded successfully.");
        } else {
            Log.e("ElectricityPredictor", "Failed to load LSTM model '" + LSTM_MODEL_FILE + "'.");
        }

        lrKerasInterpreter = loadTFLiteModel(LR_KERAS_MODEL_FILE);
        if (lrKerasInterpreter != null) {
            Log.i("ElectricityPredictor", "Keras Linear Regression model '" + LR_KERAS_MODEL_FILE + "' loaded successfully.");
        } else {
            Log.e("ElectricityPredictor", "Failed to load Keras Linear Regression model '" + LR_KERAS_MODEL_FILE + "'.");
        }

        loadScalerParams(SCALER_PARAMS_FILE);
    }

    private Interpreter loadTFLiteModel(String modelFileName) {
        try {
            return new Interpreter(loadModelFile(modelFileName));
        } catch (IOException e) {
            Log.e("ElectricityPredictor", "Error loading TFLite model '" + modelFileName + "': " + e.getMessage(), e);
            return null;
        }
    }

    private ByteBuffer loadModelFile(String modelFileName) throws IOException {
        String modelPath = modelFileName;
        FileInputStream inputStream = new FileInputStream(modelPath);
        FileChannel fileChannel = inputStream.getChannel();
        long startOffset = 0;
        long declaredLength = fileChannel.size();
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength);
    }

    private void loadScalerParams(String fileName) {
        try {
            String jsonString = readAssetFile(fileName);
            JSONObject jsonObject = new JSONObject(jsonString);
            JSONArray scaleJsonArray = jsonObject.getJSONArray("scale_");
            JSONArray minJsonArray = jsonObject.getJSONArray("min_");

            scalerScale = new float[scaleJsonArray.length()];
            scalerMin = new float[minJsonArray.length()];

            for (int i = 0; i < scaleJsonArray.length(); i++) {
                scalerScale[i] = (float) scaleJsonArray.getDouble(i);
            }
            for (int i = 0; i < minJsonArray.length(); i++) {
                scalerMin[i] = (float) minJsonArray.getDouble(i);
            }

        } catch (Exception e) {
            Log.e("ElectricityPredictor", "Error loading scaler params '" + fileName + "': " + e.getMessage(), e);
            scalerScale = null;
            scalerMin = null;
        }
    }

    private String readAssetFile(String fileName) throws IOException {
        return context.getAssets().open(fileName).bufferedReader().use(BufferedReader::readText);
    }

    private float scaleUsage(float rawValue) {
        if (scalerScale == null || scalerMin == null || scalerScale.length == 0) {
            Log.w("ElectricityPredictor", "Scaler not initialized, returning raw value for scaling.");
            return rawValue;
        }
        return rawValue * scalerScale[0] + scalerMin[0];
    }

    private float inverseScaleUsage(float scaledValue) {
        if (scalerScale == null || scalerMin == null || scalerScale.length == 0 || Math.abs(scalerScale[0]) < FLOAT_COMPARISON_THRESHOLD) {
            Log.w("ElectricityPredictor", "Scaler not initialized or scale is zero, returning scaled value for inverse scaling.");
            return scaledValue;
        }
        return (scaledValue - scalerMin[0]) / scalerScale[0];
    }

    public PredictionResult predictFutureUsage(List<ElectricityRecord> records, ModelType modelTypeToUse) {
        if (records.isEmpty()) {
            return new PredictionResult(null, Collections.emptyList(), 0.0f, "没有历史数据");
        }

        List<ElectricityRecord> sortedRecords = records.stream()
                .sorted((a, b) -> a.getTimestamp().compareTo(b.getTimestamp()))
                .collect(Collectors.toList());

        double currentBalance = sortedRecords.get(sortedRecords.size() - 1).getBalance();

        if (sortedRecords.size() < DEFAULT_INPUT_LENGTH) {
            Log.w("ElectricityPredictor", "历史数据不足 (需要 " + DEFAULT_INPUT_LENGTH + ", 现有 " + sortedRecords.size() + ")，尝试简单预测。");
            if (sortedRecords.size() >= 2) {
                double historicalDailyUsage = calculateHistoricalDailyUsage(sortedRecords);
                return simpleLinearPrediction(currentBalance, historicalDailyUsage);
            } else {
                return new PredictionResult(null, Collections.emptyList(), 0.0f, "历史数据不足以进行任何预测");
            }
        }

        List<Float> initialRawHistory = sortedRecords.subList(
                Math.max(0, sortedRecords.size() - DEFAULT_INPUT_LENGTH),
                sortedRecords.size()
        ).stream()
                .map(record -> (float) record.getChange())
                .collect(Collectors.toList());

        List<Float> initialScaledHistory = initialRawHistory.stream()
                .map(this::scaleUsage)
                .collect(Collectors.toList());

        Interpreter activeInterpreter;
        String modelNameForLog;

        switch (modelTypeToUse) {
            case LSTM:
                activeInterpreter = lstmInterpreter;
                modelNameForLog = "LSTM";
                break;
            case LINEAR_REGRESSION_KERAS:
                activeInterpreter = lrKerasInterpreter;
                modelNameForLog = "Keras LR";
                break;
            case SIMPLE_LINEAR:
                Log.i("ElectricityPredictor", "Using simple linear prediction as requested.");
                double historicalDailyUsage = calculateHistoricalDailyUsage(sortedRecords);
                return simpleLinearPrediction(currentBalance, historicalDailyUsage);
            default:
                throw new IllegalArgumentException("Unknown model type: " + modelTypeToUse);
        }

        if (activeInterpreter == null) {
            Log.e("ElectricityPredictor", modelNameForLog + " interpreter is null. Falling back to simple linear prediction.");
            double historicalDailyUsage = calculateHistoricalDailyUsage(sortedRecords);
            return simpleLinearPrediction(currentBalance, historicalDailyUsage);
        }

        List<DailyPrediction> dailyPredictions = new ArrayList<>();
        double tempRemainingBalance = currentBalance;
        LocalDateTime today = LocalDateTime.now();
        List<Float> currentScaledHistory = new ArrayList<>(initialScaledHistory);

        try {
            for (int dayOffset = 0; dayOffset < PREDICTION_HORIZON_DAYS; dayOffset++) {
                ByteBuffer inputBuffer = ByteBuffer.allocateDirect(DEFAULT_INPUT_LENGTH * 4)
                        .order(ByteOrder.nativeOrder());
                for (float value : currentScaledHistory) {
                    inputBuffer.putFloat(value);
                }
                inputBuffer.rewind();

                ByteBuffer outputBuffer = ByteBuffer.allocateDirect(MODEL_OUTPUT_STEPS * 4)
                        .order(ByteOrder.nativeOrder());

                activeInterpreter.run(inputBuffer, outputBuffer);
                outputBuffer.rewind();

                float predictedScaledUsage = outputBuffer.getFloat();
                float predictedRawUsage = inverseScaleUsage(predictedScaledUsage);

                tempRemainingBalance -= predictedRawUsage;
                dailyPredictions.add(new DailyPrediction(
                        today.plusDays(dayOffset + 1),
                        Math.max(0.0, predictedRawUsage),
                        tempRemainingBalance
                ));

                if (currentScaledHistory.size() > 0) {
                    currentScaledHistory.remove(0);
                }
                currentScaledHistory.add(predictedScaledUsage);

                if (tempRemainingBalance <= FLOAT_COMPARISON_THRESHOLD) {
                    tempRemainingBalance = 0.0;
                    for (int remainingDayOffset = dayOffset + 1; remainingDayOffset < PREDICTION_HORIZON_DAYS; remainingDayOffset++) {
                        dailyPredictions.add(new DailyPrediction(
                                today.plusDays(remainingDayOffset + 1),
                                0.0,
                                0.0
                        ));
                    }
                    break;
                }
            }

            double historicalDailyUsage = calculateHistoricalDailyUsage(sortedRecords);
            Integer daysUntilEmptyByModel = calculateDaysUntilEmpty(dailyPredictions, currentBalance, historicalDailyUsage);

            return new PredictionResult(
                    daysUntilEmptyByModel,
                    dailyPredictions,
                    calculateConfidence(dailyPredictions, historicalDailyUsage),
                    null
            );

        } catch (Exception e) {
            Log.e("ElectricityPredictor", "Error running TFLite inference with " + modelNameForLog + ": " + e.getMessage(), e);
            double historicalDailyUsage = calculateHistoricalDailyUsage(sortedRecords);
            return simpleLinearPrediction(currentBalance, historicalDailyUsage, "模型(" + modelNameForLog + ")预测失败: " + e.getLocalizedMessage());
        }
    }

    private Integer calculateDaysUntilEmpty(List<DailyPrediction> predictions, double currentBalance, double historicalDailyUsage) {
        if (predictions.isEmpty()) return null;

        boolean allPredictedUsageIsZero = predictions.stream()
                .allMatch(p -> p.getPredictedUsage() <= FLOAT_COMPARISON_THRESHOLD);

        if (allPredictedUsageIsZero && currentBalance > FLOAT_COMPARISON_THRESHOLD) {
            return null;
        }

        double balanceForDaysCalc = currentBalance;
        int days = 0;

        for (DailyPrediction prediction : predictions) {
            if (balanceForDaysCalc >= prediction.getPredictedUsage()) {
                balanceForDaysCalc -= prediction.getPredictedUsage();
                days++;
            } else {
                if (prediction.getPredictedUsage() > FLOAT_COMPARISON_THRESHOLD) {
                    days += (int) (balanceForDaysCalc / prediction.getPredictedUsage());
                }
                balanceForDaysCalc = 0.0;
                break;
            }
        }

        if (days == predictions.size() && balanceForDaysCalc > FLOAT_COMPARISON_THRESHOLD) {
            double lastPredictedUsage = predictions.get(predictions.size() - 1).getPredictedUsage();
            if (lastPredictedUsage > FLOAT_COMPARISON_THRESHOLD) {
                days += (int) (balanceForDaysCalc / lastPredictedUsage);
            } else if (historicalDailyUsage > FLOAT_COMPARISON_THRESHOLD) {
                days += (int) (balanceForDaysCalc / historicalDailyUsage);
            }
        }

        return days;
    }

    private double calculateHistoricalDailyUsage(List<ElectricityRecord> records) {
        if (records.size() < 2) return 0.0;

        double totalNetUsage = records.stream()
                .mapToDouble(ElectricityRecord::getChange)
                .sum();

        LocalDateTime firstTimestamp = records.get(0).getTimestamp();
        LocalDateTime lastTimestamp = records.get(records.size() - 1).getTimestamp();
        double daysBetween = ChronoUnit.DAYS.between(firstTimestamp, lastTimestamp);

        if (daysBetween > 0 && totalNetUsage > 0) {
            return totalNetUsage / daysBetween;
        } else if (records.size() == 1 && records.get(0).getChange() > 0) {
            return records.get(0).getChange();
        } else {
            return 0.0;
        }
    }

    private PredictionResult simpleLinearPrediction(double currentBalance, double historicalDailyUsage, String customError) {
        if (historicalDailyUsage <= FLOAT_COMPARISON_THRESHOLD && currentBalance > FLOAT_COMPARISON_THRESHOLD) {
            List<DailyPrediction> zeroPredictions = new ArrayList<>();
            for (int i = 0; i < PREDICTION_HORIZON_DAYS; i++) {
                zeroPredictions.add(new DailyPrediction(
                        LocalDateTime.now().plusDays(i + 1),
                        0.0,
                        currentBalance
                ));
            }
            return new PredictionResult(null, zeroPredictions, 0.5f, customError != null ? customError : "日均用电为零但有余额");
        }

        if (historicalDailyUsage <= FLOAT_COMPARISON_THRESHOLD && currentBalance <= FLOAT_COMPARISON_THRESHOLD) {
            return new PredictionResult(0, Collections.emptyList(), 1.0f, customError);
        }

        List<DailyPrediction> dailyPredictions = new ArrayList<>();
        double tempRemainingBalance = currentBalance;
        LocalDateTime today = LocalDateTime.now();

        for (int i = 0; i < PREDICTION_HORIZON_DAYS; i++) {
            double predictedUsageForDay = historicalDailyUsage;
            tempRemainingBalance -= predictedUsageForDay;
            if (tempRemainingBalance < 0) tempRemainingBalance = 0.0;
            dailyPredictions.add(new DailyPrediction(
                    today.plusDays(i + 1),
                    Math.max(0.0, predictedUsageForDay),
                    tempRemainingBalance
            ));
            if (tempRemainingBalance == 0.0) break;
        }

        while (dailyPredictions.size() < PREDICTION_HORIZON_DAYS && tempRemainingBalance == 0.0) {
            dailyPredictions.add(new DailyPrediction(
                    today.plusDays(dailyPredictions.size() + 1),
                    0.0,
                    0.0
            ));
        }

        Integer daysUntilEmpty = historicalDailyUsage > FLOAT_COMPARISON_THRESHOLD
                ? (int) (currentBalance / historicalDailyUsage)
                : null;

        return new PredictionResult(daysUntilEmpty, dailyPredictions, 0.7f, customError);
    }

    private float calculateConfidence(List<DailyPrediction> modelPredictions, double historicalDailyUsage) {
        if (modelPredictions.isEmpty()) return 0.0f;

        double avgModelPredictedUsage = modelPredictions.stream()
                .mapToDouble(DailyPrediction::getPredictedUsage)
                .average()
                .orElse(0.0);

        if (historicalDailyUsage <= FLOAT_COMPARISON_THRESHOLD) {
            return Math.abs(avgModelPredictedUsage) <= FLOAT_COMPARISON_THRESHOLD ? 0.9f : 0.1f;
        }

        double difference = Math.abs(avgModelPredictedUsage - historicalDailyUsage);
        double normalizedDifference = difference / historicalDailyUsage;
        return (float) Math.max(0.0, Math.min(1.0, 1.0 - normalizedDifference));
    }

    public void close() {
        if (lstmInterpreter != null) {
            lstmInterpreter.close();
            lstmInterpreter = null;
        }
        if (lrKerasInterpreter != null) {
            lrKerasInterpreter.close();
            lrKerasInterpreter = null;
        }
        Log.i("ElectricityPredictor", "All TFLite interpreters closed.");
    }

    public enum ModelType {
        LSTM,
        LINEAR_REGRESSION_KERAS,
        SIMPLE_LINEAR
    }

    public static class PredictionResult {
        private final Integer daysUntilEmpty;
        private final List<DailyPrediction> predictions;
        private final float confidence;
        private final String error;

        public PredictionResult(Integer daysUntilEmpty, List<DailyPrediction> predictions,
                              float confidence, String error) {
            this.daysUntilEmpty = daysUntilEmpty;
            this.predictions = predictions;
            this.confidence = confidence;
            this.error = error;
        }

        public Integer getDaysUntilEmpty() { return daysUntilEmpty; }
        public List<DailyPrediction> getPredictions() { return predictions; }
        public float getConfidence() { return confidence; }
        public String getError() { return error; }
    }

    public static class DailyPrediction {
        private final LocalDateTime date;
        private final double predictedUsage;
        private final double remainingBalance;

        public DailyPrediction(LocalDateTime date, double predictedUsage, double remainingBalance) {
            this.date = date;
            this.predictedUsage = predictedUsage;
            this.remainingBalance = remainingBalance;
        }

        public LocalDateTime getDate() { return date; }
        public double getPredictedUsage() { return predictedUsage; }
        public double getRemainingBalance() { return remainingBalance; }
    }
} 

File: app/src/main/java/com/kirawii/thunderswufe/ml/ElectricityPredictor.kt
package com.kirawii.thunderswufe.ml

import android.content.Context
import android.util.Log
import com.kirawii.thunderswufe.data.database.ElectricityRecord
import org.json.JSONObject
import org.tensorflow.lite.Interpreter
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import kotlin.math.max

enum class ModelType {
    LINEAR_REGRESSION_KERAS,
    SIMPLE_LINEAR
}

class ElectricityPredictor(private val context: Context) {
    private var lrKerasInterpreter: Interpreter? = null

    // 模型文件名常量
    companion object {
        private const val LR_KERAS_MODEL_FILE = "linear_regression_keras.tflite"
        private const val SCALER_PARAMS_FILE = "scaler_params.json"
        private const val DEFAULT_INPUT_LENGTH = 7
        private const val MODEL_OUTPUT_STEPS = 1
        private const val PREDICTION_HORIZON_DAYS = 7
    }

    private var scalerScale: FloatArray? = null
    private var scalerMin: FloatArray? = null

    private val floatComparisonThreshold = 1e-6f

    init {
        loadAllModelsAndParams()
    }

    private fun loadAllModelsAndParams() {
        lrKerasInterpreter = loadTFLiteModel(LR_KERAS_MODEL_FILE)
        if (lrKerasInterpreter != null) {
            Log.i("ElectricityPredictor", "Keras Linear Regression model '$LR_KERAS_MODEL_FILE' loaded successfully.")
        } else {
            Log.e("ElectricityPredictor", "Failed to load Keras Linear Regression model '$LR_KERAS_MODEL_FILE'.")
        }

        loadScalerParams(SCALER_PARAMS_FILE)
        if (scalerScale == null || scalerMin == null) {
            Log.e("ElectricityPredictor", "Failed to load scaler parameters from '$SCALER_PARAMS_FILE'. Predictions might be inaccurate.")
        }
    }

    private fun loadTFLiteModel(modelFileName: String): Interpreter? {
        return try {
            context.assets.open(modelFileName).use { inputStream ->
                val bytes = inputStream.readBytes()
                val byteBuffer = ByteBuffer.allocateDirect(bytes.size)
                byteBuffer.order(ByteOrder.nativeOrder())
                byteBuffer.put(bytes)
                byteBuffer.rewind()
                Interpreter(byteBuffer)
            }
        } catch (e: Exception) {
            Log.e("ElectricityPredictor", "TFLite模型 '$modelFileName' 加载失败：${e.message}", e)
            null
        }
    }

    private fun loadScalerParams(fileName: String) {
        try {
            val jsonString = context.assets.open(fileName).bufferedReader().use { it.readText() }
            val jsonObject = JSONObject(jsonString)
            val scaleJsonArray = jsonObject.getJSONArray("scale_")
            val minJsonArray = jsonObject.getJSONArray("min_")

            scalerScale = FloatArray(scaleJsonArray.length()) { i -> scaleJsonArray.getDouble(i).toFloat() }
            scalerMin = FloatArray(minJsonArray.length()) { i -> minJsonArray.getDouble(i).toFloat() }
            Log.i("ElectricityPredictor", "Scaler params loaded: scale=${scalerScale?.contentToString()}, min=${scalerMin?.contentToString()}")

        } catch (e: Exception) {
            Log.e("ElectricityPredictor", "Error loading scaler params '$fileName': ${e.message}", e)
            scalerScale = null
            scalerMin = null
        }
    }

    private fun scaleUsage(rawValue: Float): Float {
        if (scalerScale == null || scalerMin == null || scalerScale!!.isEmpty()) {
            Log.w("ElectricityPredictor", "Scaler not initialized, returning raw value for scaling.")
            return rawValue
        }
        return rawValue * scalerScale!![0] + scalerMin!![0]
    }

    private fun inverseScaleUsage(scaledValue: Float): Float {
        if (scalerScale == null || scalerMin == null || scalerScale!!.isEmpty() || scalerScale!![0] == 0f) {
            Log.w("ElectricityPredictor", "Scaler not initialized or scale is zero, returning scaled value for inverse scaling.")
            return scaledValue
        }
        return (scaledValue - scalerMin!![0]) / scalerScale!![0]
    }

    fun predictFutureUsage(
        records: List<ElectricityRecord>,
        modelTypeToUse: ModelType = ModelType.LINEAR_REGRESSION_KERAS
    ): PredictionResult {
        if (records.isEmpty()) {
            return PredictionResult(null, emptyList(), 0.0f, "没有历史数据")
        }

        val sortedRecords = records.sortedBy { it.timestamp }
        val currentBalance = sortedRecords.last().balance

        if (sortedRecords.size < DEFAULT_INPUT_LENGTH) {
            Log.w("ElectricityPredictor", "历史数据不足 (需要 $DEFAULT_INPUT_LENGTH, 现有 ${sortedRecords.size})，无法进行预测。")
            return PredictionResult(null, emptyList(), 0.0f, "历史数据不足以进行任何预测")
        }

        val initialRawHistory = sortedRecords
            .takeLast(DEFAULT_INPUT_LENGTH)
            .map { it.change.toFloat() }

        if (initialRawHistory.size < DEFAULT_INPUT_LENGTH) {
            Log.e("ElectricityPredictor", "Logic error: Not enough data for initial history after filtering.")
            return PredictionResult(null, emptyList(), 0.0f, "内部错误：准备历史数据失败")
        }
        val initialScaledHistory = initialRawHistory.map { scaleUsage(it) }.toMutableList()

        val activeInterpreter = lrKerasInterpreter
        val modelNameForLog = "Keras LR"

        if (activeInterpreter == null) {
            val modelFileName = LR_KERAS_MODEL_FILE
            Log.e("ElectricityPredictor", "$modelNameForLog interpreter is null. 无法进行预测。请检查 assets 目录下的 $modelFileName 是否存在且为有效 TFLite 模型。")
            return PredictionResult(null, emptyList(), 0.0f, "$modelNameForLog interpreter is null. 无法进行预测。请检查 assets 目录下的 $modelFileName 是否存在且为有效 TFLite 模型。")
        }
        if (scalerScale == null || scalerMin == null){
            Log.e("ElectricityPredictor", "Scaler params are null. Predictions will be inaccurate. 无法进行预测。")
            return PredictionResult(null, emptyList(), 0.0f, "Scaler params are null. 无法进行预测。")
        }

        val dailyPredictions = mutableListOf<DailyPrediction>()
        var tempRemainingBalance = currentBalance
        val today = LocalDateTime.now()
        var currentScaledHistory = initialScaledHistory.toMutableList() // 可变副本

        try {
            var dayOffset = 0
            while (tempRemainingBalance > floatComparisonThreshold && dayOffset < 365) { // 最多预测一年，防止死循环
                if (currentScaledHistory.size < DEFAULT_INPUT_LENGTH) {
                    Log.e("ElectricityPredictor", "History size became less than $DEFAULT_INPUT_LENGTH during prediction loop.")
                    break // Should not happen
                }
                // 准备模型输入
                val inputBuffer = ByteBuffer.allocateDirect(DEFAULT_INPUT_LENGTH * Float.SIZE_BYTES)
                    .order(ByteOrder.nativeOrder())
                currentScaledHistory.forEach { inputBuffer.putFloat(it) }
                inputBuffer.rewind()

                val outputBuffer = ByteBuffer.allocateDirect(MODEL_OUTPUT_STEPS * Float.SIZE_BYTES)
                    .order(ByteOrder.nativeOrder())

                activeInterpreter.run(inputBuffer, outputBuffer)
                outputBuffer.rewind()

                val predictedScaledUsage = outputBuffer.float // 只取一个 float
                val predictedRawUsage = inverseScaleUsage(predictedScaledUsage)

                tempRemainingBalance -= predictedRawUsage
                dailyPredictions.add(
                    DailyPrediction(
                        date = today.plusDays(dayOffset.toLong() + 1),
                        predictedUsage = max(0.0f, predictedRawUsage).toDouble(),
                        remainingBalance = tempRemainingBalance
                    )
                )
                // 更新预测输入，移除最早的记录，添加新的预测结果
                currentScaledHistory.removeAt(0)
                currentScaledHistory.add(scaleUsage(predictedRawUsage.toFloat()))

                if (tempRemainingBalance <= floatComparisonThreshold) {
                    tempRemainingBalance = 0.0
                    break
                }
                dayOffset += 1 // 每次只步进一天

            }

            val historicalDailyUsage = calculateHistoricalDailyUsage(sortedRecords) // 用于置信度和可能的耗尽估算
            var daysUntilEmptyByModel: Int? = null
            var balanceForDaysCalc = currentBalance
            var days = 0
            var allPredictedUsageIsZero = true
            for (prediction in dailyPredictions) {
                if (prediction.predictedUsage > floatComparisonThreshold) {
                    allPredictedUsageIsZero = false
                }
                if (balanceForDaysCalc >= prediction.predictedUsage) {
                    balanceForDaysCalc -= prediction.predictedUsage
                    days++
                } else {
                    if (prediction.predictedUsage > floatComparisonThreshold) {
                        days += (balanceForDaysCalc / prediction.predictedUsage).toInt()
                    }
                    balanceForDaysCalc = 0.0
                    break
                }
            }

            if (days == dailyPredictions.size && balanceForDaysCalc > floatComparisonThreshold) {
                val lastPredictedUsage = dailyPredictions.lastOrNull()?.predictedUsage ?: historicalDailyUsage
                if (lastPredictedUsage > floatComparisonThreshold) {
                    days += (balanceForDaysCalc / lastPredictedUsage).toInt()
                } else if (historicalDailyUsage > floatComparisonThreshold){
                    days += (balanceForDaysCalc / historicalDailyUsage).toInt()
                } else {

                }
            }
            daysUntilEmptyByModel = if (allPredictedUsageIsZero && currentBalance > floatComparisonThreshold) null else days

            return PredictionResult(
                daysUntilEmpty = daysUntilEmptyByModel,
                predictions = dailyPredictions,
                confidence = calculateConfidence(dailyPredictions, historicalDailyUsage),
                error = null
            )

        }catch(e: Exception){
            Log.e("ElectricityPredictor", "预测异常: ${e.message}", e)
            return PredictionResult(null, emptyList(), 0.0f, "预测发生异常: ${e.localizedMessage}")
        }

    }

    private fun calculateHistoricalDailyUsage(records: List<ElectricityRecord>): Double {
        if (records.size < 2) return 0.0
        val sortedForCalc = if (records.first().timestamp.isAfter(records.last().timestamp)) {
            records.sortedBy { it.timestamp }
        } else { records }

        val totalNetUsage = sortedForCalc.map { it.change }.sum().toDouble()
        val firstTimestamp = sortedForCalc.first().timestamp
        val lastTimestamp = sortedForCalc.last().timestamp
        val daysBetween = ChronoUnit.DAYS.between(firstTimestamp, lastTimestamp).toDouble()

        return if (daysBetween > 0 && totalNetUsage > 0) {
            totalNetUsage / daysBetween
        } else if (sortedForCalc.size == 1 && sortedForCalc.first().change > 0) {
            sortedForCalc.first().change.toDouble()
        }
        else { 0.0 }
    }

    private fun calculateConfidence(modelPredictions: List<DailyPrediction>, historicalDailyUsage: Double): Float {
        if (modelPredictions.isEmpty()) return 0.0f
        val avgModelPredictedUsage = modelPredictions.map { it.predictedUsage }.average()
        if (historicalDailyUsage <= floatComparisonThreshold) {
            return if (kotlin.math.abs(avgModelPredictedUsage) <= floatComparisonThreshold) 0.9f else 0.1f
        }
        val difference = kotlin.math.abs(avgModelPredictedUsage - historicalDailyUsage)
        val normalizedDifference = difference / historicalDailyUsage
        return (1.0f - normalizedDifference.toFloat()).coerceIn(0.0f, 1.0f)
    }

    fun close() {
        lrKerasInterpreter?.close()
        lrKerasInterpreter = null
        Log.i("ElectricityPredictor", "All TFLite interpreters closed.")
    }
}

data class PredictionResult(
    val daysUntilEmpty: Int?,
    val predictions: List<DailyPrediction>,
    val confidence: Float,
    val error: String?
)

data class DailyPrediction(
    val date: LocalDateTime,
    val predictedUsage: Double,
    val remainingBalance: Double
)